File: src/main/java/editor/gui/filter/editor/OptionsFilterPanel.java
package editor.gui.filter.editor;

import java.awt.BorderLayout;
import java.awt.Color;
import java.awt.Dimension;
import java.awt.Font;
import java.awt.GridLayout;
import java.awt.Point;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

import javax.swing.BorderFactory;
import javax.swing.BoxLayout;
import javax.swing.JComboBox;
import javax.swing.JLabel;
import javax.swing.JPanel;
import javax.swing.JScrollPane;
import javax.swing.SwingUtilities;
import javax.swing.plaf.basic.BasicComboPopup;

import editor.filter.Filter;
import editor.filter.FilterAttribute;
import editor.filter.leaf.FilterLeaf;
import editor.filter.leaf.options.OptionsFilter;
import editor.gui.generic.ButtonScrollPane;
import editor.gui.generic.ComboBoxPanel;
import editor.gui.generic.ScrollablePanel;
import editor.util.Containment;
import editor.util.MouseListenerFactory;
import editor.util.PopupMenuListenerFactory;
import editor.util.UnicodeSymbols;

/**
 * This class represents a panel that corresponds to a filter that groups
 * cards according to a characteristic that takes on distinct values.  Unlike
 * other filter panels, which can be switched among different types of filters
 * as long as they are the same class, OptionsFilterPanel cannot.
 *
 * @param <T> Type that the options presented have
 * @author Alec Roelke
 */
@SuppressWarnings("serial")
public class OptionsFilterPanel<T> extends FilterEditorPanel<OptionsFilter<T>>
{
    /**
     * Maximum width for combo boxes.  Sizes of the drop-down menus remain
     * unaffected.
     */
    private static final int MAX_COMBO_WIDTH = 100;

    /**
     * Set containment combo box.
     */
    private ComboBoxPanel<Containment> contain;
    /**
     * List of options that are available to choose from.
     */
    private T[] options;
    /**
     * List of boxes displaying the currently-selected options.
     */
    private List<JComboBox<T>> optionsBoxes;
    /**
     * Panel displaying the combo boxes to be used to choose
     * options.
     */
    private ScrollablePanel optionsPanel;
    /**
     * Type of filter this OptionsFilterPanel edits.
     */
    private FilterAttribute type;

    /**
     * Create a new OptionsFilterPanel using the given filter to initialize its
     * fields and the given array to specify the set of options to choose from.
     *
     * @param f filter to use for initialization
     * @param t list of options to choose from
     */
    public OptionsFilterPanel(OptionsFilter<T> f, T[] t)
    {
        this(f.type(), t);
        setContents(f);
    }

    /**
     * Create a new OptionsFilterPanel.
     *
     * @param t type of the new OptionsFilterPanel
     * @param o list of options to choose from
     */
    public OptionsFilterPanel(FilterAttribute t, T[] o)
    {
        super();
        setLayout(new BorderLayout());

        type = t;
        options = o;
        optionsBoxes = new ArrayList<>();

        contain = new ComboBoxPanel<>(Containment.values());
        add(contain, BorderLayout.WEST);

        optionsPanel = new ScrollablePanel(ScrollablePanel.TRACK_HEIGHT);
        optionsPanel.setLayout(new BoxLayout(optionsPanel, BoxLayout.X_AXIS));
        ButtonScrollPane optionsPane = new ButtonScrollPane(optionsPanel);
        optionsPane.setBorder(BorderFactory.createEmptyBorder());
        add(optionsPane, BorderLayout.CENTER);
    }

    /**
     * Add a new combo box for an additional option.
     *
     * @param value initial value of the new combo box.
     */
    private void addItem(T value)
    {
        JPanel boxPanel = new JPanel(new BorderLayout());
        var box = new JComboBox<>(options);
        box.addPopupMenuListener(PopupMenuListenerFactory.createVisibleListener((e) -> {
            if (options.length > 0)
            {
                Object child = box.getAccessibleContext().getAccessibleChild(0);
                if (child instanceof BasicComboPopup)
                    SwingUtilities.invokeLater(() -> {
                        BasicComboPopup popup = (BasicComboPopup)child;
                        JScrollPane scrollPane = (JScrollPane)SwingUtilities.getAncestorOfClass(JScrollPane.class, popup.getList());

                        int popupWidth = popup.getList().getPreferredSize().width +
                                (options.length > box.getMaximumRowCount() ? scrollPane.getVerticalScrollBar().getPreferredSize().width : 0);
                        scrollPane.setPreferredSize(new Dimension(Math.max(popupWidth, scrollPane.getPreferredSize().width), scrollPane.getPreferredSize().height));
                        scrollPane.setMaximumSize(scrollPane.getPreferredSize());
                        Point location = box.getLocationOnScreen();
                        popup.setLocation(location.x, location.y + box.getHeight() - 1);
                        popup.setLocation(location.x, location.y + box.getHeight());
                    });
            }
        }));
        box.setPreferredSize(new Dimension(MAX_COMBO_WIDTH, box.getPreferredSize().height));

        boxPanel.add(box, BorderLayout.CENTER);
        optionsBoxes.add(box);
        box.setSelectedItem(value);

        JPanel buttonPanel = new JPanel(new GridLayout(2, 1, 0, 0));
        JLabel addButton = new JLabel("+", JLabel.CENTER);
        Font buttonFont = new Font(addButton.getFont().getFontName(), Font.PLAIN, addButton.getFont().getSize() - 2);
        addButton.setAlignmentX(CENTER_ALIGNMENT);
        addButton.setFont(buttonFont);
        addButton.addMouseListener(MouseListenerFactory.createPressListener((e) -> {
            addItem(options[0]);
            optionsPanel.revalidate();
        }));
        JLabel removeButton = new JLabel(String.valueOf(UnicodeSymbols.MULTIPLY), JLabel.CENTER);
        removeButton.setForeground(Color.RED);
        removeButton.setAlignmentX(CENTER_ALIGNMENT);
        removeButton.setFont(buttonFont);
        removeButton.addMouseListener(MouseListenerFactory.createPressListener((e) -> {
            if (optionsBoxes.size() > 1)
            {
                optionsPanel.remove(boxPanel);
                optionsBoxes.remove(box);
                optionsPanel.revalidate();
            }
        }));
        buttonPanel.add(removeButton);
        buttonPanel.add(addButton);
        boxPanel.add(buttonPanel, BorderLayout.EAST);

        optionsPanel.add(boxPanel);
    }

    @Override
    public Filter filter()
    {
        @SuppressWarnings("unchecked")
        var filter = (OptionsFilter<T>)FilterAttribute.createFilter(type);
        filter.contain = contain.getSelectedItem();
        filter.selected = optionsBoxes.stream().map((b) -> b.getItemAt(b.getSelectedIndex())).collect(Collectors.toSet());
        return filter;
    }

    /**
     * {@inheritDoc}
     *
     * @throws IllegalArgumentException if the given filter is not the same
     *                                  type as this OptionsFilterPanel or isn't even an {@link OptionsFilter}
     */
    @SuppressWarnings("unchecked")
    @Override
    public void setContents(FilterLeaf<?> filter) throws IllegalArgumentException
    {
        if (filter instanceof OptionsFilter && filter.type().equals(type))
            setContents((OptionsFilter<T>)filter);
        else if (filter instanceof OptionsFilter)
            throw new IllegalArgumentException("Options filter type " + filter.type() + " does not match type " + type);
        else
            throw new IllegalArgumentException("Illegal options filter " + filter.type());
    }

    /**
     * {@inheritDoc}
     *
     * @throws IllegalArgumentException if the given filter is of the wrong type
     */
    @Override
    public void setContents(OptionsFilter<T> filter) throws IllegalArgumentException
    {
        if (filter.type().equals(type))
        {
            contain.setSelectedItem(filter.contain);
            if (options.length == 0)
                contain.setVisible(false);
            optionsBoxes.clear();
            optionsPanel.removeAll();
            if (filter.selected.isEmpty() && options.length > 0)
                addItem(options[0]);
            else
                for (T selected : filter.selected)
                    addItem(selected);
        }
        else
            throw new IllegalArgumentException("Options filter type " + filter.type() + " does not match type " + type);
    }
}

File: src/main/scala/editor/gui/filter/editor/OptionsFilterPanel.scala
package editor.gui.filter.editor

import editor.database.attributes.CardAttribute
import editor.filter.leaf.FilterLeaf
import editor.filter.leaf.options.OptionsFilter
import editor.gui.filter.FilterSelectorPanel
import editor.gui.generic.ButtonScrollPane
import editor.gui.generic.ComboBoxPanel
import editor.gui.generic.ScrollablePanel
import editor.util.Containment
import editor.util.MouseListenerFactory
import editor.util.PopupMenuListenerFactory
import editor.util.UnicodeSymbols

import java.awt.BorderLayout
import java.awt.Color
import java.awt.Component
import java.awt.Dimension
import java.awt.Font
import java.awt.GridLayout
import javax.swing.BorderFactory
import javax.swing.BoxLayout
import javax.swing.JComboBox
import javax.swing.JLabel
import javax.swing.JPanel
import javax.swing.JScrollPane
import javax.swing.SwingConstants
import javax.swing.SwingUtilities
import javax.swing.plaf.basic.BasicComboPopup
import scala.jdk.CollectionConverters._
import scala.reflect.ClassTag

/**
 * Convenience constructors for [[OptionsFilterPanel]].
 * @author Alec Roelke
 */
object OptionsFilterPanel {
  /**
   * Create a new [[OptionsFilterPanel]] that filters the given attribute using the given options.
   * 
   * @param attribute card attribute to filter
   * @param options values of the attribute that can be chosen
   * @return an [[OptionsFilterPanel]] for a filter of the given attribute with the given possible values
   */
  def apply[T <: AnyRef : ClassTag](attribute: CardAttribute, options: Seq[T], selector: FilterSelectorPanel) = new OptionsFilterPanel(attribute, options, selector)

  /**
   * Create a new [[OptionsFilterPanel]] with the given filter and possible options to choose from. The attribute
   * used for filtering is inferred form the filter.
   * 
   * @param filter filter used to infer the attribute and pre-populate the panel
   * @param options values of the attribute that can be chosen
   * @return an [[OptionsFilterPanel]] populated with the contents of the filter
   */
  def apply[T <: AnyRef : ClassTag](filter: OptionsFilter[T], options: Seq[T], selector: FilterSelectorPanel) = {
    val panel = new OptionsFilterPanel(filter.attribute, options, selector)
    panel.setContents(filter)
    panel
  }
}

/**
 * A filter editor panel that is used to customize a filter that selects one or more items from a number of
 * predefined options. Each selected option is a combo box with all of the options and a pair of small buttons
 * that allows for adding another selection or removing an existing one. If enough boxes are added, arrows on
 * the left and right side will become enabled to allow for scrolling.
 * 
 * @constructor create a new options filter panel for filtering a particular attribute using a particular set of
 * options
 * @param attribute card attribute to filter for
   @param options options that can be chosen for the filter value
 * 
 * @author Alec Roelke
 */
class OptionsFilterPanel[T <: AnyRef : ClassTag](protected override val attribute: CardAttribute, options: Seq[T], selector: FilterSelectorPanel) extends FilterEditorPanel[OptionsFilter[T]] {
  private val MaxComboWidth = 100

  setLayout(BorderLayout())

  private val boxes = collection.mutable.Buffer[JComboBox[T]]()

  private val contain = ComboBoxPanel(Containment.values.toArray)
  add(contain, BorderLayout.WEST)

  private val optionsPanel = ScrollablePanel(ScrollablePanel.TrackHeight)
  optionsPanel.setLayout(BoxLayout(optionsPanel, BoxLayout.X_AXIS))
  private val optionsPane = ButtonScrollPane(optionsPanel)
  optionsPane.setBorder(BorderFactory.createEmptyBorder)
  add(optionsPane, BorderLayout.CENTER)

  private def +=(value: T): Unit = {
    val boxPanel = JPanel(BorderLayout())
    val box = JComboBox(options.toArray)
    box.addPopupMenuListener(PopupMenuListenerFactory.createPopupListener(visible = _ => if (!options.isEmpty) {
      box.getAccessibleContext.getAccessibleChild(0) match {
        case popup: BasicComboPopup => SwingUtilities.invokeLater(() => SwingUtilities.getAncestorOfClass(classOf[JScrollPane], popup.getList) match {
          case scrollPane: JScrollPane =>
            val popupWidth = popup.getList.getPreferredSize.width + (if (options.size > box.getMaximumRowCount) scrollPane.getVerticalScrollBar.getPreferredSize.width else 0)
            scrollPane.setPreferredSize(Dimension(math.max(popupWidth, scrollPane.getPreferredSize.width), scrollPane.getPreferredSize.height))
            scrollPane.setMaximumSize(scrollPane.getPreferredSize)
            val location = box.getLocationOnScreen
            popup.setLocation(location.x, location.y + box.getHeight - 1)
            popup.setLocation(location.x, location.y + box.getHeight)
        })
      }
    }))
    box.setPreferredSize(Dimension(MaxComboWidth, box.getPreferredSize.height))

    boxPanel.add(box, BorderLayout.CENTER)
    boxes += box
    box.setSelectedItem(value)

    val buttonPanel = JPanel(GridLayout(2, 1, 0, 0))
    val addButton = JLabel("+", SwingConstants.CENTER)
    val buttonFont = Font(addButton.getFont.getFontName, Font.PLAIN, addButton.getFont.getSize - 2)
    addButton.setAlignmentX(Component.CENTER_ALIGNMENT)
    addButton.setFont(buttonFont)
    addButton.addMouseListener(MouseListenerFactory.createMouseListener(pressed = _ => {
      this += options(0)
      optionsPanel.revalidate()
    }))
    val removeButton = JLabel(UnicodeSymbols.MULTIPLY.toString, SwingConstants.CENTER)
    removeButton.setForeground(Color.RED)
    removeButton.setAlignmentX(Component.CENTER_ALIGNMENT)
    removeButton.setFont(buttonFont)
    removeButton.addMouseListener(MouseListenerFactory.createMouseListener(pressed = _ => if (!boxes.isEmpty) {
      optionsPanel.remove(boxPanel)
      boxes -= box
      optionsPanel.revalidate()
    }))
    buttonPanel.add(removeButton)
    buttonPanel.add(addButton)
    boxPanel.add(buttonPanel, BorderLayout.EAST)

    optionsPanel.add(boxPanel)
  }

  override def filter = CardAttribute.createFilter(attribute) match {
    case of: OptionsFilter[T] =>
      of.faces = selector.faces
      of.contain = contain.getSelectedItem
      of.selected = boxes.map((b) => b.getItemAt(b.getSelectedIndex)).toSet
      of
  }

  override def setFields(filter: OptionsFilter[T]) = if (filter.attribute == attribute) {
    contain.setSelectedItem(filter.contain)
    if (options.isEmpty)
      contain.setVisible(false)
    boxes.clear()
    optionsPanel.removeAll()
    if (filter.selected.isEmpty && !options.isEmpty)
      this += options(0)
    else
      filter.selected.foreach(this += _)
  } else throw IllegalArgumentException(s"${filter.attribute} is not a $attribute filter")
}